File: packages/cli-js/src/utils/shell.ts
/**
 * Shell / process utilities (run claude, codex, etc.)
 */

import { execSync, spawnSync } from "child_process";
import { existsSync } from "fs";
import { join } from "path";

const isWindows = process.platform === "win32";

/**
 * Common npm global install locations (for sandboxed environments where PATH may be restricted).
 * Used by resolveCommandPathWithNpmFallback.
 */
export function getNpmGlobalExeCandidates(cmd: string): string[] {
  if (isWindows) {
    const appData = process.env.APPDATA;
    const localAppData = process.env.LOCALAPPDATA;
    const candidates = [
      appData && join(appData, "npm", `${cmd}.cmd`),
      appData && join(appData, "npm", cmd),
      localAppData && join(localAppData, "npm", `${cmd}.cmd`),
      localAppData && join(localAppData, "npm", cmd),
    ].filter((p): p is string => Boolean(p));
    return candidates;
  }
  const home = process.env.HOME;
  if (!home) return [];
  return [
    join(home, ".npm-global", "bin", cmd),
    join(home, ".local", "bin", cmd),
    join(home, "npm", "bin", cmd),
    "/usr/local/bin/" + cmd,
  ];
}

/**
 * Resolve full path to command: try PATH first, then common npm global locations.
 * Works in sandboxed environments (e.g. Cursor agent) where PATH may not include npm global bin.
 */
export function resolveCommandPathWithNpmFallback(cmd: string): string | null {
  const fromPath = resolveCommandPath(cmd);
  if (fromPath) return fromPath;
  const candidate = getNpmGlobalExeCandidates(cmd).find((p) => existsSync(p));
  return candidate ?? null;
}

export function runCommand(cmd: string, args: string[], cwd?: string): string {
  const full = [cmd, ...args].join(" ");
  return execSync(full, { encoding: "utf-8", cwd }).trim();
}

/**
 * Run a command with stdio inherited so output is visible in the terminal. Returns true if exit code is 0.
 */
export function runCommandLive(cmd: string, args: string[], cwd?: string): boolean {
  const result = spawnSync(cmd, args, { stdio: "inherit", cwd, shell: true });
  return result.status === 0;
}

export function hasCommand(cmd: string): boolean {
  try {
    const check = isWindows ? `where ${cmd}` : `which ${cmd}`;
    execSync(check, { encoding: "utf-8", stdio: "pipe" });
    return true;
  } catch {
    return false;
  }
}

/**
 * Resolve full path to command executable (for spawning without shell so arguments are preserved).
 * On Windows, prefers .cmd so we get the real npm global wrapper (e.g. codex.cmd) not the extensionless entry.
 */
export function resolveCommandPath(cmd: string): string | null {
  try {
    const check = isWindows ? `where ${cmd}` : `which ${cmd}`;
    const out = execSync(check, { encoding: "utf-8", stdio: "pipe" });
    const lines = out.split(/[\r\n]+/).map((s) => s.trim()).filter(Boolean);
    if (isWindows && lines.length > 0) {
      const preferred = lines.find((p) => p.toLowerCase().endsWith(".cmd"));
      const chosen = preferred ?? lines.find((p) => existsSync(p)) ?? lines[0];
      return chosen || null;
    }
    return lines[0] || null;
  } catch {
    return null;
  }
}

File: packages/cli-py/planforge/utils/shell.py
"""Shell / process utilities (run claude, codex, etc.)."""

import os
import subprocess
import shutil
from pathlib import Path


def run_command(cmd: str, args: list[str], cwd: str | None = None) -> str:
    result = subprocess.run(
        [cmd, *args],
        capture_output=True,
        text=True,
        cwd=cwd,
    )
    result.check_returncode()
    return (result.stdout or "").strip()


def has_command(cmd: str) -> bool:
    return shutil.which(cmd) is not None


def get_npm_global_exe_candidates(cmd: str) -> list[str]:
    """Common npm global install locations (for sandboxed environments where PATH may be restricted)."""
    if os.name == "nt":
        candidates = []
        app_data = os.environ.get("APPDATA")
        local_app_data = os.environ.get("LOCALAPPDATA")
        if app_data:
            candidates.append(str(Path(app_data) / "npm" / f"{cmd}.cmd"))
            candidates.append(str(Path(app_data) / "npm" / cmd))
        if local_app_data:
            candidates.append(str(Path(local_app_data) / "npm" / f"{cmd}.cmd"))
            candidates.append(str(Path(local_app_data) / "npm" / cmd))
        return candidates
    home = os.environ.get("HOME")
    if not home:
        return []
    return [
        str(Path(home) / ".npm-global" / "bin" / cmd),
        str(Path(home) / ".local" / "bin" / cmd),
        str(Path(home) / "npm" / "bin" / cmd),
        f"/usr/local/bin/{cmd}",
    ]


def resolve_command_path_with_npm_fallback(cmd: str) -> str | None:
    """Resolve full path to command: try PATH first, then common npm global locations.
    Works in sandboxed environments (e.g. Cursor agent) where PATH may not include npm global bin.
    """
    exe = shutil.which(cmd)
    if exe:
        return exe
    for path in get_npm_global_exe_candidates(cmd):
        if os.path.isfile(path):
            return path
    return None
